### ISSUE TITLE
Combine changelog and changelog_eng

### ISSUE THREAD

**Issue by MKadaner** [FarGroup/FarManager#253]
Starting at 64c221090e27a3adf8e9ed82a5f3fa7d6f785d22, file `changelog_eng` is considered the first-class official Far Manager change list equal to the Russian version at `changelog`. Yay!

Now, every developer have to update two files with the summary of their changes. It's not that hard, but still taxing. How about combining the files together? There is obvious advantage of lower maintenance. Not only one file need to be updated instead of two, but some (many?) records, like `Refactoring` or `Correction of` can be made in English only; hopefully, such simple comments can easily be understood by any reader.

Here is a proposal of how the combined file might look like: [far/changelog.md](https://github.com/MKadaner/FarManager/blob/mzk-changelog/far/changelog.md). And here is the comment by @alabuzhev to start the discussion:

> Some points to consider:
> - The files are huge: github already can't render changelog and changelog_eng is close to that. Merging them into even huger one won't improve the situation, so any formatting doesn't really make sense.
> - It might be easier to update the merged version, but I'm not sure about reading it - a reader will have to jump over several lines over and over again.
>
> _Originally posted by @alabuzhev in https://github.com/FarGroup/FarManager/pull/252_

File size is indeed an issue. However, it is not strictly tied to combing the files together. They are already large; something need to be done, combining or not. A possible solution is to split the file(s) at strategic places, like it was done with famous [Ralf Brown's Interrupt List](https://www.cs.cmu.edu/~ralf/files.html). Then, the pieces could be linked together with ordinary hyperlinks. Global search over the entire contents would suffer, especially online. How often global search is required, especially online?

Readability is a real problem. Do you have any notion of how often people read change log? Long ago, to decide whether I wanted to update Far at certain point, I tried to read change log and glean what new nice features were implemented or annoying bug were fixed. It was rather difficult anyway, mostly for three reasons:

1. No highlighting. The headings and comments are not easily distinguishable, at least for my poor eyesight. This could be neatly fixed with markdown.
2. Lots of "uninteresting" records like `Refactoring` or `Correction of`.
3. Sometimes, it's difficult to understand whether a new feature considered completed or still in experimental phase; there is a record saying "added such and such feature," then several records about corrections and continuations, but usually there is no final record saying it's production ready.

(The latter two issues are traditionally solved by maintaining release notes, but writing them is a task on its own. I do not think this overhead is manageable for such a small team as Far developers'.)

Now, I simply upgrade Far once so often without bothering to look at the change log. My point here is that improving writing experience in this case may be more important than improving readability (as much as I hate saying such things). Again, with proper use of markdown and smaller files allowing to read them online, bilingual clutter might be acceptable.

**Comment by johnd0e** [comment]
> My point here is that improving writing experience in this case may be more important than improving readability

We could improve both: write single source file, and transform it into two during build process.

**Comment by shmuz** [comment]
I think we can stop maintaining Russian version of the change log. Even if our English is not perfect it will suffice.

**Comment by alabuzhev** [comment]
Michael, thanks for writing this down.

> Starting at [64c2210](https://github.com/FarGroup/FarManager/commit/64c221090e27a3adf8e9ed82a5f3fa7d6f785d22), file `changelog_eng` is considered the first-class official Far Manager change list equal to the Russian version at `changelog`. Yay!
> 
> Now, every developer have to update two files with the summary of their changes. It's not that hard, but still taxing.

I wish I could persuade others to *actually* do that.

> How about combining the files together? There is obvious advantage of lower maintenance. Not only one file need to be updated instead of two, but some (many?) records, like `Refactoring` or `Correction of` can be made in English only; hopefully, such simple comments can easily be understood by any reader.

Indeed, such purely technical notes don't need no translation. In fact, they don't even need a changelog entry, a commit description would suffice.

> File size is indeed an issue.

It's not an issue right now, because changelogs are plain text files, and the fact that github can't render them fancily changes nothing. However, if we convert them to md, that will be an issue, because what's the point of having md if it can't be viewed as md? (unless the reader/writer has md-enabled desktop software of course)

> A possible solution is to split the file(s) at strategic places

The thing is, we don't have that many "strategic" places. The last one (2.x -> 3.x) was 9.5 years and 800 Kb ago.

> Readability is a real problem. Do you have any notion of how often people read change log?

Judging by the comments in the forum - they don't. I hope I'm wrong though.

> 1. No highlighting. The headings and comments are not easily distinguishable, at least for my poor eyesight. This could be neatly fixed with markdown.

Good point. However, I'm not sure that [your example](https://github.com/MKadaner/FarManager/blob/mzk-changelog/far/changelog.md) is more readable - it makes headers (names, timestamps, build numbers) more pronounced, but that's the least useful information in the log.   

> 3. Sometimes, it's difficult to understand whether a new feature considered completed or still in experimental phase; there is a record saying "added such and such feature," then several records about corrections and continuations, but usually there is no final record saying it's production ready.

Unless it is explicitly marked as "experimental" - it is considered ready straight away (at least in the eye of the author).

> (The latter two issues are traditionally solved by maintaining release notes, but writing them is a task on its own. I do not think this overhead is manageable for such a small team as Far developers'.)

Release notes imply releases, and we don't have releases in the conventional sense. The reasons for that is a long story, too long and too sad to be posted here.

> Now, I simply upgrade Far once so often without bothering to look at the change log.

That's usually ok, but beware - sometimes we drop really experimental stuff into the cauldron. It's better to at least look for road blocks like "warning" and "experimental".

> My point here is that improving writing experience in this case may be more important than improving readability (as much as I hate saying such things).

I agree - even a badly written text is better than no text at all.

To summarise:
Changelogs as we know them are not a problem per se, be it a single file or two. They are consequences of the current development process, which is a consequence of the certain events, decisions and limitations from days of long ago. It could be improved, but that requires work, and "no work is less work than some work".

**Comment by MKadaner** [comment]
Thank you for the thoughtful comments, Alex.

> unless the reader/writer has md-enabled desktop software of course

Far editor _is_ an md-enabled software. :)

> we don't have that many "strategic" places

"Strategic" is not the same as "meaningful" :)

> it makes headers (names, timestamps, build numbers) more pronounced, but that's the least useful information in the log

It works differently, at least for me. I can as easily skip over the headings as over the small print. It's a matter of mental tuning. «Здесь играем, здесь не играем, а здесь рыбу заворачивали…» Well, you know the tune.

> Changelogs as we know them are not a problem per se, be it a single file or two. ... It could be improved, but that requires work, and "no work is less work than some work".

I think that with a very limited effort the situation can be improved, even if not dramatically.

Please look at [this variant](https://github.com/MKadaner/FarManager/blob/mzk-changelog/far/changelog_05.md). With a little more effort I can remove verbatim duplicates as well as Russian variants of trivially translatable records. I did not create a formal PR because I do not want to appear pushy. If you think the approach is viable, please let me know what other tweaks are worthwhile, and I'll create a PR. If it feels too cumbersome, let's just drop the matter while no harm is done.

**Comment by alabuzhev** [comment]
> Please look at this variant.

This one is a little confusing, at least for me. Separators visually take precedence and appear as "entry separators", rather than "translation separators within the same entry", so I can't stop seeing entries with sudden headers in the middle.

> I think that with a very limited effort the situation can be improved

When I said "it could be improved" I meant the development process, but that's a different topic (some thoughts [in the forum](https://forum.farmanager.com/viewtopic.php?p=152164#p152164)) .
Speaking of effort, how much would it take to:
- add nice and useful md formatting where it's needed (looking at your example - not that much)?
- avoid accidental, intrusive and annoying md formatting where it's not needed (unknown, potentially noticeable)?

Example:
Turning entry headers into md headers is easy - I just add `###` to a macro that generates them and forget about it.
However, remembering to escape special symbols like `>` or `` ` `` and all md quirks like swallowing new lines, two spaces == \<br> etc. and always previewing it somewhere (because it never looks the way you expected on the first try) is *the last thing I want* when I've just spent some time working on some sensible changes and simply want to describe them and call it a day.

> I do not want to appear pushy

No worries, you're certainly not. It's an interesting topic to discuss and to explore. Even if we don't reach a consensus - changelogs are already better than they were :)

Personally, I like Shmuel's suggestion above - English only will suffice. I'd like to go even further and say "commit descriptions will suffice", but our users probably won't appreciate that.

**Comment by MKadaner** [comment]
OK, it looks like markdown does not work; line breaks are especially tricky. FarColorer would nicely highlight un-escaped special characters but combining lines into paragraphs and gratuitous `<br>`s is a nuisance. Forget about markdown.

The choice of separators was not particularly good. My previous example was to demonstrate overall layout rather than specific formatting. This is easy to fix. I was thinking about bilingual layout for some time but could not come up with anything better than two sections, one after the other. Two columns would be the best, but it will not work in a text file.

I agree that simplifying development process is a very important, maybe the most important, consideration here. I learned from my own experience that maintaining two parallel changelogs is a nuisance and unnecessary burden on a developer who wants to quickly finalize a substantial change and pronounce a victory. It is exacerbated by writing commit descriptions, but that is a different story, probably coupled with "stable builds."

Personally, I like Shmuel's suggestion very much. I hesitated to second it because my opinion on this topic is deeply biased. With my reformatting, I am suggesting a way to transition from two changelogs to a single combined one (split into chunks because of tools' limitations) with bilingual or unilingual descriptions as developers see fit (English preferred). Formatting is secondary (thought important). If you think this goal is not worth it, let's drop this issue.

With this purpose in mind, please look at [the next variant](https://github.com/MKadaner/FarManager/blob/mzk-changelog/far/changelog_05). If it's the right direction, we can think of chunk numbering scheme (e.g. changelog, changelog_04, changelog_03, etc.), formatting, and the comments which could be made unilingual from the beginning (like "Refactoring").

**Comment by MKadaner** [comment]
@alabuzhev gentle ping.

**Comment by alabuzhev** [comment]
@MKadaner, this looks interesting, although the "· · ·" separator is a little confusing - it looks like the ellipsis (omission) rather than the dinkus (division). Also, it could be better to put English section first. 

Re chunk numbering scheme - I'm not sure that we still need it if we're not switching to md, as it was mostly related to rendering issues. What other tools' limitations do we have?

P.S. I wanted to respond earlier but got distracted, please feel free to ping sooner if needed.

**Comment by MKadaner** [comment]
Please look at [this version](https://github.com/MKadaner/FarManager/blob/mzk-changelog/misc/changelog/changelog_00). Besides tweaking formatting, I folded some bilingual records into English only. The list of patterns which I considered foldable is in [this script](https://github.com/MKadaner/FarManager/blob/mzk-changelog/misc/changelog/changelog.linq) (starting on the line 25). More patterns should be added. I think this is the most useful piece of this proposal.

As for the tools, GitHub can render this file in raw mode only; no formatting and no blame. Probably not big deal for this contents. Atlassian Source Tree does not show diff for large files, but the limit is configurable, and I almost never use AST anyway. Also, it took about 3 seconds to fully load the file in browser (Edge, pretty fast connection, reasonably fast machine). How often do people need to look at the bottom of the changelog? Approximately never, I guess. All in all, chunking is the least important feature. One can always cut off the tail and safe it separately when one gets too annoyed by the size of the file.

**Comment by alabuzhev** [comment]
Looks like we're getting there :) The idea with folding trivial records is neat and the script is impressive 👍. This way we can also gradually migrate to English-only version while retaining the localised records for archaeological purposes.

I think raw mode is fine - there's no formatting anyway and blaming the changelog is hardly ever necessary.

**Comment by MKadaner** [comment]
Thank you. I will finish with the folding and ask you to cross-check the result. When everything looks fine, I'll create pull request.

**Comment by MKadaner** [comment]
I created draft PR #323. Please review.

**Comment by MKadaner** [comment]
@alabuzhev could you please take a look at #323? Alternatively, you can review the files in the [clone repo](https://github.com/MKadaner/FarManager/tree/mzk-changelog/misc/changelog).

**Comment by alabuzhev** [comment]
@MKadaner It looks good, thank you.

I have one concern though - I noticed that people still tend to ignore `changelog_eng` and only update `changelog`.
It's not a big deal now - it's quite obvious that the recent entries are absent and it's easy to sync and add them retrospectively.
Supposedly with a single file approach the chance that these absent entries will stay unnoticed is higher.

**Comment by MKadaner** [comment]
@alabuzhev How about this [macro](https://github.com/MKadaner/FarManager/blob/mzk-changelog/misc/changelog/ChangelogCheckEnglish.lua)?

Granted, one (you) will need to run it manually from time to time, but isn't it even easier than keeping an eye on missing entries in `changelog_eng`?

**Comment by alabuzhev** [comment]
That might work, thank you.
